### PR TITLE
expression: fix EvaluateExprWithNull when meet cast(#3873)

### LINKED ISSUE #3873: error `FUNCTION cast does not exist` happens when doing out join simplification
Please answer these questions before submitting your issue. Thanks!

out join simplication: https://dev.mysql.com/doc/refman/5.7/en/outer-join-simplification.html

1. What did you do?
``` sql
create table tb1(id int);
create table tb2(id int);
select * from tb1 left join tb2 on tb1.id=tb2.id where tb2.id>1.0;
```

2. What did you expect to see?
``` sql
0 rows in set (0.00 sec)
```

3. What did you see instead?
``` sql
ERROR 1305 (42000): FUNCTION cast does not exist
```

4. What version of TiDB are you using (`tidb-server -V`)?
```
Git Commit Hash: d99f2974fb21b1215fed10c2c6dcfbc24073f706
UTC Build Time:  2017-07-12 03:36:02
```

### PR DESCRIPTION
**Pull request by XuHuaiyu**
`cast` is built in `NewCastFunc` but not NewFunction,
this commit is a tradeoff,
I'll move the building of cast to NewFunction.

fix #3873 
PTAL 
@zz-jason @hanfei1991

### CODE REVIEW

File: expression/expression.go [hunk 1]
```diff
@@ -443,9 +443,14 @@ func EvaluateExprWithNull(ctx context.Context, schema *Schema, expr Expression)
 				return nil, errors.Trace(err)
 			}
 		}
-		newFunc, err := NewFunction(ctx, x.FuncName.L, types.NewFieldType(mysql.TypeTiny), args...)
-		if err != nil {
-			return nil, errors.Trace(err)
+		var newFunc Expression
+		if x.FuncName.O == ast.Cast {
```
Comment (zz-jason) [inline expression/expression.go]: How about do this check in `NewFunction` ?

Comment (XuHuaiyu) [inline expression/expression.go]: good idea

Comment (hanfei1991) [inline expression/expression.go]: Why not check funcName.L

**Comment by XuHuaiyu** [comment]
PTAL  again @zz-jason @hanfei1991

**Pull request merged**